File: src/mobile/ios/XerahSMobile/Features/Settings/CustomUploaderConfigScreen.swift
//
//  CustomUploaderConfigScreen.swift
//  XerahS Mobile (Swift)
//
//  XerahS - The Avalonia UI implementation of ShareX
//  Copyright (c) 2007-2026 ShareX Team
//
//  This program is free software; you can redistribute it and/or
//  modify it under the terms of the GNU General Public License
//  as published by the Free Software Foundation; either version 2
//  of the License, or (at your option) any later version.
//
//  This program is distributed in the hope that it will be useful,
//  but WITHOUT ANY WARRANTY; without even the implied warranty of
//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//  GNU General Public License for more details.
//
//  You should have received a copy of the GNU General Public License
//  along with this program; if not, write to the Free Software
//  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
//
//  Optionally you can also view the license at <http://www.gnu.org/licenses/>.
//

import SwiftUI
import UIKit

struct CustomUploaderConfigScreen: View {
    @ObservedObject var viewModel: CustomUploaderConfigViewModel

    var body: some View {
        List {
            if let status = viewModel.statusMessage, !status.isEmpty {
                Section {
                    HStack(spacing: 12) {
                        Image(systemName: viewModel.isStatusError ? "exclamationmark.triangle.fill" : "checkmark.circle.fill")
                            .foregroundStyle(viewModel.isStatusError ? .red : .green)

                        Text(status)
                            .foregroundStyle(viewModel.isStatusError ? .red : .primary)

                        Spacer()

                        if viewModel.isStatusError, viewModel.importErrorDetails != nil {
                            Button {
                                viewModel.copyImportErrorToClipboard()
                            } label: {
                                Label("Copy Error", systemImage: "doc.on.doc")
                            }
                            .labelStyle(.iconOnly)
                        }
                    }
                }
            }

            Section {
                if viewModel.uploaders.isEmpty {
                    ContentUnavailableView(
                        "No Custom Uploaders",
                        systemImage: "curlybraces",
                        description: Text("Import a `.sxcu` from the clipboard or create one manually.")
                    )
                } else {
                    ForEach(viewModel.uploaders) { entry in
                        CustomUploaderRow(
                            entry: entry,
                            onEdit: { viewModel.edit(entry) },
                            onCopy: { viewModel.copySxcuToClipboard(entry) },
                            onDelete: { viewModel.delete(entry) }
                        )
                    }
                }
            } footer: {
                Text("Custom uploaders use the same request, body, and response shape as desktop.")
            }
        }
        .navigationTitle("Custom Uploader")
        .toolbar {
            ToolbarItemGroup(placement: .topBarTrailing) {
                Button {
                    viewModel.importFromClipboard()
                } label: {
                    Label("Import Clipboard", systemImage: "doc.on.clipboard")
                }

                Button {
                    viewModel.addNew()
                } label: {
                    Label("New Custom Uploader", systemImage: "plus")
                }
            }
        }
        .onAppear { viewModel.refresh() }
        .sheet(item: $viewModel.editingEntry) { entry in
            CustomUploaderEditSheet(
                entry: entry,
                onDismiss: { viewModel.cancelEdit() },
                onSave: { viewModel.saveEdit($0) }
            )
        }
    }
}

private struct CustomUploaderRow: View {
    let entry: CustomUploaderEntry
    var onEdit: () -> Void
    var onCopy: () -> Void
    var onDelete: () -> Void

    var body: some View {
        HStack(alignment: .top, spacing: 12) {
            Image(systemName: "curlybraces")
                .foregroundStyle(.blue)

            VStack(alignment: .leading, spacing: 4) {
                Text(entry.displayName)
                    .font(.headline)
                Text(entry.requestUrl.isEmpty ? "No URL configured" : entry.requestUrl)
                    .font(.caption)
                    .foregroundStyle(.secondary)
                    .lineLimit(2)
                Text("\(entry.requestMethod.rawValue) - \(entry.bodyType.rawValue) - \(entry.destinationType)")
                    .font(.caption2.weight(.semibold))
                    .foregroundStyle(.secondary)
            }

            Spacer()

            Menu {
                Button(action: onEdit) {
                    Label("Edit", systemImage: "pencil")
                }
                Button(action: onCopy) {
                    Label("Copy .sxcu", systemImage: "doc.on.doc")
                }
                Button(role: .destructive, action: onDelete) {
                    Label("Delete", systemImage: "trash")
                }
            } label: {
                Label("Actions", systemImage: "ellipsis.circle")
                    .labelStyle(.iconOnly)
            }
        }
    }
}

private struct CustomUploaderEditSheet: View {
    let entry: CustomUploaderEntry
    var onDismiss: () -> Void
    var onSave: (CustomUploaderEntry) -> Void

    @State private var name: String = ""
    @State private var destinationType: String = ""
    @State private var requestMethod: CustomUploaderRequestMethod = .POST
    @State private var requestUrl: String = ""
    @State private var bodyType: CustomUploaderBodyType = .multipartFormData
    @State private var fileFormName: String = "file"
    @State private var parametersText: String = ""
    @State private var headersText: String = ""
    @State private var argumentsText: String = ""
    @State private var dataText: String = ""
    @State private var urlText: String = ""
    @State private var deletionUrlText: String = ""
    @State private var errorMessageText: String = ""

    var body: some View {
        NavigationStack {
            Form {
                Section("Basic") {
                    TextField("Name", text: $name)
                    TextField("Destination Type", text: $destinationType)
                        .textInputAutocapitalization(.never)
                    Picker("Request Method", selection: $requestMethod) {
                        ForEach(CustomUploaderRequestMethod.allCases, id: \.self) { method in
                            Text(method.rawValue).tag(method)
                        }
                    }
                    TextField("Request URL", text: $requestUrl, axis: .vertical)
                        .textInputAutocapitalization(.never)
                        .keyboardType(.URL)
                        .lineLimit(2...4)
                    Picker("Body Type", selection: $bodyType) {
                        ForEach(CustomUploaderBodyType.allCases, id: \.self) { type in
                            Text(type.rawValue).tag(type)
                        }
                    }
                    TextField("File Form Name", text: $fileFormName)
                        .textInputAutocapitalization(.never)
                }

                Section("Parameters") {
                    KeyValueTextEditor(text: $parametersText, prompt: "key=value")
                }

                Section("Headers") {
                    KeyValueTextEditor(text: $headersText, prompt: "Header=Value")
                }

                Section("Arguments") {
                    KeyValueTextEditor(text: $argumentsText, prompt: "field=value")
                }

                Section("Body Data") {
                    TextEditor(text: $dataText)
                        .frame(minHeight: 120)
                        .font(.system(.body, design: .monospaced))
                }

                Section("Response Parsing") {
                    TextField("URL template", text: $urlText, axis: .vertical)
                        .textInputAutocapitalization(.never)
                        .lineLimit(2...4)
                    TextField("Deletion URL template", text: $deletionUrlText, axis: .vertical)
                        .textInputAutocapitalization(.never)
                        .lineLimit(2...4)
                    TextField("Error message template", text: $errorMessageText, axis: .vertical)
                        .textInputAutocapitalization(.never)
                        .lineLimit(2...4)
                }
            }
            .navigationTitle(entry.id.isEmpty ? "New .sxcu" : "Edit .sxcu")
            .navigationBarTitleDisplayMode(.inline)
            .toolbar {
                ToolbarItem(placement: .cancellationAction) {
                    Button("Cancel", action: onDismiss)
                }
                ToolbarItem(placement: .confirmationAction) {
                    Button("Save") {
                        var updated = entry
                        updated.name = name.trimmingCharacters(in: .whitespacesAndNewlines)
                        updated.destinationType = destinationType.trimmingCharacters(in: .whitespacesAndNewlines)
                        updated.requestMethod = requestMethod
                        updated.requestUrl = requestUrl.trimmingCharacters(in: .whitespacesAndNewlines)
                        updated.bodyType = bodyType
                        updated.fileFormName = fileFormName.trimmingCharacters(in: .whitespacesAndNewlines)
                        updated.parameters = KeyValueCodec.decode(parametersText)
                        updated.headers = KeyValueCodec.decode(headersText)
                        updated.arguments = KeyValueCodec.decode(argumentsText)
                        updated.data = dataText
                        updated.url = urlText.trimmingCharacters(in: .whitespacesAndNewlines)
                        updated.deletionUrl = deletionUrlText.trimmingCharacters(in: .whitespacesAndNewlines)
                        updated.errorMessage = errorMessageText.trimmingCharacters(in: .whitespacesAndNewlines)
                        onSave(updated)
                    }
                }
            }
            .onAppear {
                name = entry.name
                destinationType = entry.destinationType
                requestMethod = entry.requestMethod
                requestUrl = entry.requestUrl
                bodyType = entry.bodyType
                fileFormName = entry.fileFormName
                parametersText = KeyValueCodec.encode(entry.parameters)
                headersText = KeyValueCodec.encode(entry.headers)
                argumentsText = KeyValueCodec.encode(entry.arguments)
                dataText = entry.data
                urlText = entry.url.isEmpty ? entry.legacyUrlExpression : entry.url
                deletionUrlText = entry.deletionUrl
                errorMessageText = entry.errorMessage
            }
        }
    }
}

private struct KeyValueTextEditor: View {
    @Binding var text: String
    let prompt: String

    var body: some View {
        ZStack(alignment: .topLeading) {
            if text.isEmpty {
                Text(prompt)
                    .foregroundStyle(.tertiary)
                    .padding(.top, 8)
                    .padding(.leading, 4)
            }

            TextEditor(text: $text)
                .frame(minHeight: 110)
                .font(.system(.body, design: .monospaced))
                .textInputAutocapitalization(.never)
        }
    }
}

final class CustomUploaderConfigViewModel: ObservableObject {
    @Published var uploaders: [CustomUploaderEntry] = []
    @Published var editingEntry: CustomUploaderEntry?
    @Published var statusMessage: String?
    @Published var isStatusError: Bool = false
    @Published var importErrorDetails: String?

    private let settingsRepository: SettingsRepository
    private let decoder = JSONDecoder()
    private let encoder = JSONEncoder()

    init(settingsRepository: SettingsRepository) {
        self.settingsRepository = settingsRepository
        encoder.outputFormatting = [.prettyPrinted, .sortedKeys]
    }

    func refresh() {
        uploaders = settingsRepository.loadCustomUploaders()
    }

    func addNew() {
        editingEntry = CustomUploaderEntry(
            id: "custom_\(UUID().uuidString.prefix(8))",
            name: "New Uploader",
            destinationType: "FileUploader",
            requestMethod: .POST,
            requestUrl: "",
            bodyType: .multipartFormData,
            fileFormName: "file"
        )
    }

    func edit(_ entry: CustomUploaderEntry) {
        editingEntry = entry
    }

    func saveEdit(_ entry: CustomUploaderEntry) {
        var list = uploaders
        if let idx = list.firstIndex(where: { $0.id == entry.id }) {
            list[idx] = entry
        } else {
            list.append(entry)
        }
        settingsRepository.saveCustomUploaders(list)
        uploaders = list
        editingEntry = nil
        setStatus("Saved \(entry.displayName).")
    }

    func cancelEdit() {
        editingEntry = nil
    }

    func delete(_ entry: CustomUploaderEntry) {
        uploaders = uploaders.filter { $0.id != entry.id }
        settingsRepository.saveCustomUploaders(uploaders)
        setStatus("Deleted \(entry.displayName).")
    }

    func importFromClipboard() {
        guard let payload = clipboardPayload() else {
            setStatus(
                "Clipboard does not contain .sxcu JSON or a readable .sxcu file path.",
                isError: true,
                details: clipboardDiagnostics(payload: nil, cleanedString: UIPasteboard.general.string)
            )
            return
        }

        do {
            let definition = try decoder.decode(SxcuDefinition.self, from: payload)
            let entry = CustomUploaderEntry.from(sxcu: definition)
            var list = uploaders
            list.append(entry)
            settingsRepository.saveCustomUploaders(list)
            if settingsRepository.getDefaultDestinationInstanceId() == nil {
                settingsRepository.setDefaultDestinationInstanceId(entry.id)
            }
            uploaders = list
            setStatus("Imported \(entry.displayName) from clipboard.")
        } catch {
            let clipboardString = UIPasteboard.general.string?.trimmingCharacters(in: .whitespacesAndNewlines)
            setStatus(
                "Failed to import .sxcu from clipboard: \(error.localizedDescription)",
                isError: true,
                details: clipboardDiagnostics(payload: payload, cleanedString: clipboardString, error: error)
            )
        }
    }

    func copyImportErrorToClipboard() {
        guard let importErrorDetails else { return }
        UIPasteboard.general.string = importErrorDetails
        setStatus("Copied import error details to clipboard.")
    }

    func copySxcuToClipboard(_ entry: CustomUploaderEntry) {
        do {
            let data = try encoder.encode(entry.toSxcuDefinition())
            UIPasteboard.general.string = String(decoding: data, as: UTF8.self)
            setStatus("Copied \(entry.displayName) as .sxcu JSON.")
        } catch {
            setStatus("Failed to encode .sxcu JSON: \(error.localizedDescription)", isError: true)
        }
    }

    private func clipboardPayload() -> Data? {
        if let string = UIPasteboard.general.string?.trimmingCharacters(in: .whitespacesAndNewlines), !string.isEmpty {
            let cleaned = stripMarkdownCodeFence(from: string)
            if cleaned.hasPrefix("file://"), let url = URL(string: cleaned) {
                return try? Data(contentsOf: url)
            }
            if cleaned.hasSuffix(".sxcu"), FileManager.default.fileExists(atPath: cleaned) {
                return try? Data(contentsOf: URL(fileURLWithPath: cleaned))
            }
            return cleaned.data(using: .utf8)
        }

        if let url = UIPasteboard.general.url {
            return try? Data(contentsOf: url)
        }

        return nil
    }

    private func stripMarkdownCodeFence(from input: String) -> String {
        guard input.hasPrefix("```") else { return input }
        let lines = input.components(separatedBy: .newlines)
        guard lines.count >= 3 else { return input }
        return lines.dropFirst().dropLast().joined(separator: "\n")
    }

    private func clipboardDiagnostics(payload: Data?, cleanedString: String?, error: Error? = nil) -> String {
        let cleaned = cleanedString.map(stripMarkdownCodeFence(from:))?.trimmingCharacters(in: .whitespacesAndNewlines)
        let preview = cleaned.map { String($0.prefix(400)) }
        let payloadText = payload.flatMap { String(data: $0, encoding: .utf8) }
        let decodedJsonKeys: String? = {
            guard
                let payload,
                let object = try? JSONSerialization.jsonObject(with: payload) as? [String: Any]
            else { return nil }
            return object.keys.sorted().joined(separator: ", ")
        }()

        var lines: [String] = [
            "SXCU clipboard import failed",
            "Timestamp: \(ISO8601DateFormatter().string(from: Date()))",
            "Clipboard has string: \((UIPasteboard.general.string?.isEmpty == false) ? "true" : "false")",
            "Clipboard has URL: \(UIPasteboard.general.url != nil ? "true" : "false")",
            "Payload bytes: \(payload?.count ?? 0)"
        ]

        if let error {
            lines.append("Error: \(String(describing: error))")
        }
        if let decodedJsonKeys, !decodedJsonKeys.isEmpty {
            lines.append("Decoded top-level keys: \(decodedJsonKeys)")
        }
        if let preview, !preview.isEmpty {
            lines.append("Clipboard preview:")
            lines.append(preview)
        } else if let payloadText, !payloadText.isEmpty {
            lines.append("Payload preview:")
            lines.append(String(payloadText.prefix(400)))
        }

        return lines.joined(separator: "\n")
    }

    private func setStatus(_ message: String, isError: Bool = false, details: String? = nil) {
        statusMessage = message
        isStatusError = isError
        importErrorDetails = isError ? details : nil
    }
}

private enum KeyValueCodec {
    static func encode(_ value: [String: String]) -> String {
        value.keys.sorted().map { key in
            "\(key)=\(value[key] ?? "")"
        }.joined(separator: "\n")
    }

    static func decode(_ text: String) -> [String: String] {
        var result: [String: String] = [:]

        for rawLine in text.components(separatedBy: .newlines) {
            let line = rawLine.trimmingCharacters(in: .whitespacesAndNewlines)
            guard !line.isEmpty else { continue }

            if let equals = line.firstIndex(of: "=") {
                let key = String(line[..<equals]).trimmingCharacters(in: .whitespacesAndNewlines)
                let value = String(line[line.index(after: equals)...]).trimmingCharacters(in: .whitespacesAndNewlines)
                if !key.isEmpty { result[key] = value }
                continue
            }

            if let colon = line.firstIndex(of: ":") {
                let key = String(line[..<colon]).trimmingCharacters(in: .whitespacesAndNewlines)
                let value = String(line[line.index(after: colon)...]).trimmingCharacters(in: .whitespacesAndNewlines)
                if !key.isEmpty { result[key] = value }
            }
        }

        return result
    }
}

File: src/mobile/android/feature/settings/src/main/java/com/getsharex/xerahs/mobile/feature/settings/CustomUploaderConfigScreen.kt
/*
 * XerahS - The Avalonia UI implementation of ShareX
 * Copyright (c) 2007-2026 ShareX Team
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 *
 * Optionally you can also view the license at <http://www.gnu.org/licenses/>.
 */

package com.getsharex.xerahs.mobile.feature.settings

import android.content.ClipData
import android.content.ClipboardManager
import android.content.Context
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.AlertDialog
import androidx.compose.material3.Button
import androidx.compose.material3.Card
import androidx.compose.material3.CardDefaults
import androidx.compose.material3.FloatingActionButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.unit.dp
import com.getsharex.xerahs.mobile.core.domain.CustomUploaderBodyType
import com.getsharex.xerahs.mobile.core.data.SettingsRepository
import com.getsharex.xerahs.mobile.core.domain.CustomUploaderEntry
import com.getsharex.xerahs.mobile.core.domain.CustomUploaderRequestMethod
import androidx.lifecycle.viewmodel.compose.viewModel

@Composable
fun CustomUploaderConfigScreen(
    settingsRepository: SettingsRepository?,
    onBack: () -> Unit
) {
    if (settingsRepository == null) {
        Button(onClick = onBack) { Text("Back") }
        return
    }
    val viewModel: CustomUploaderConfigViewModel = viewModel(
        factory = object : androidx.lifecycle.ViewModelProvider.Factory {
            @Suppress("UNCHECKED_CAST")
            override fun <T : androidx.lifecycle.ViewModel> create(modelClass: Class<T>): T =
                CustomUploaderConfigViewModel(settingsRepository) as T
        }
    )
    val uploaders by viewModel.uploaders.collectAsState()
    val editing by viewModel.editingEntry.collectAsState()
    val statusMessage by viewModel.statusMessage.collectAsState()
    val isStatusError by viewModel.isStatusError.collectAsState()
    val importErrorDetails by viewModel.importErrorDetails.collectAsState()
    val context = LocalContext.current
    val clipboardManager = context.getSystemService(Context.CLIPBOARD_SERVICE) as? ClipboardManager
    fun copyToClipboard(label: String, text: String) {
        clipboardManager?.setPrimaryClip(ClipData.newPlainText(label, text))
    }

    Column(modifier = Modifier.fillMaxSize().padding(16.dp)) {
        Row(
            modifier = Modifier.fillMaxWidth(),
            horizontalArrangement = Arrangement.SpaceBetween,
            verticalAlignment = Alignment.CenterVertically
        ) {
            Button(onClick = onBack) { Text("Back") }
            Button(
                onClick = {
                    viewModel.importFromClipboardText(clipboardManager?.primaryClip?.getItemAt(0)?.coerceToText(context)?.toString())
                }
            ) {
                Text("Import Clipboard")
            }
        }
        Spacer(modifier = Modifier.height(16.dp))
        Text(
            text = "Custom Uploader",
            style = MaterialTheme.typography.titleLarge
        )
        Text(
            text = "Import and edit .sxcu definitions using the same request, body, and response shape as desktop.",
            style = MaterialTheme.typography.bodySmall,
            color = MaterialTheme.colorScheme.onSurfaceVariant
        )
        Spacer(modifier = Modifier.height(8.dp))
        if (!statusMessage.isNullOrBlank()) {
            Card(
                modifier = Modifier.fillMaxWidth(),
                colors = CardDefaults.cardColors()
            ) {
                Column(modifier = Modifier.padding(12.dp)) {
                    Text(
                        text = statusMessage.orEmpty(),
                        color = if (isStatusError) MaterialTheme.colorScheme.error else MaterialTheme.colorScheme.onSurface,
                        style = MaterialTheme.typography.bodySmall
                    )
                    if (isStatusError && !importErrorDetails.isNullOrBlank()) {
                        OutlinedButton(
                            onClick = {
                                copyToClipboard("sxcu import error", importErrorDetails.orEmpty())
                                viewModel.markImportErrorCopied()
                            }
                        ) {
                            Text("Copy Error")
                        }
                    }
                }
            }
            Spacer(modifier = Modifier.height(8.dp))
        }
        LazyColumn(verticalArrangement = Arrangement.spacedBy(8.dp)) {
            items(uploaders, key = { it.id }) { entry ->
                Card(
                    modifier = Modifier.fillMaxWidth(),
                    colors = CardDefaults.cardColors()
                ) {
                    Row(
                        modifier = Modifier.padding(12.dp),
                        verticalAlignment = Alignment.CenterVertically
                    ) {
                        Column(modifier = Modifier.weight(1f)) {
                            Text(
                                text = entry.displayName,
                                style = MaterialTheme.typography.titleSmall
                            )
                            Text(
                                text = entry.requestUrl.ifBlank { "No URL" },
                                style = MaterialTheme.typography.bodySmall
                            )
                            Text(
                                text = "${entry.requestMethod.name} • ${entry.bodyType.name} • ${entry.destinationType}",
                                style = MaterialTheme.typography.bodySmall,
                                color = MaterialTheme.colorScheme.onSurfaceVariant
                            )
                        }
                        OutlinedButton(onClick = { viewModel.edit(entry) }) { Text("Edit") }
                        Spacer(modifier = Modifier.padding(4.dp))
                        OutlinedButton(
                            onClick = {
                                copyToClipboard("sxcu", viewModel.sxcuJson(entry))
                                viewModel.markExported(entry)
                            }
                        ) { Text("Copy .sxcu") }
                        Spacer(modifier = Modifier.padding(4.dp))
                        OutlinedButton(onClick = { viewModel.delete(entry) }) { Text("Delete") }
                    }
                }
            }
        }
        Spacer(modifier = Modifier.height(8.dp))
        FloatingActionButton(
            onClick = { viewModel.addNew() }
        ) {
            Text("Add")
        }
    }

    editing?.let { entry ->
        CustomUploaderEditDialog(
            entry = entry,
            onDismiss = { viewModel.cancelEdit() },
            onSave = { updated -> viewModel.saveEdit(updated) }
        )
    }
}

@Composable
private fun CustomUploaderEditDialog(
    entry: CustomUploaderEntry,
    onDismiss: () -> Unit,
    onSave: (CustomUploaderEntry) -> Unit
) {
    var name by remember { mutableStateOf(entry.name) }
    var destinationType by remember { mutableStateOf(entry.destinationType) }
    var requestMethod by remember { mutableStateOf(entry.requestMethod.name) }
    var requestUrl by remember { mutableStateOf(entry.requestUrl) }
    var bodyType by remember { mutableStateOf(entry.bodyType.name) }
    var fileFormName by remember { mutableStateOf(entry.fileFormName) }
    var parametersText by remember { mutableStateOf(KeyValueCodec.encode(entry.parameters)) }
    var headersText by remember { mutableStateOf(KeyValueCodec.encode(entry.headers)) }
    var argumentsText by remember { mutableStateOf(KeyValueCodec.encode(entry.arguments)) }
    var dataText by remember { mutableStateOf(entry.data) }
    var urlText by remember { mutableStateOf(entry.url.ifBlank { entry.urlExpression }) }
    var deletionUrlText by remember { mutableStateOf(entry.deletionUrl) }
    var errorMessageText by remember { mutableStateOf(entry.errorMessage) }

    AlertDialog(
        onDismissRequest = onDismiss,
        title = { Text(if (entry.id.isNotEmpty()) "Edit .sxcu" else "New .sxcu") },
        text = {
            Column(
                modifier = Modifier
                    .fillMaxWidth()
                    .verticalScroll(rememberScrollState())
            ) {
                OutlinedTextField(
                    value = name,
                    onValueChange = { name = it },
                    label = { Text("Name") },
                    modifier = Modifier.fillMaxWidth()
                )
                Spacer(modifier = Modifier.height(8.dp))
                OutlinedTextField(
                    value = destinationType,
                    onValueChange = { destinationType = it },
                    label = { Text("Destination Type") },
                    modifier = Modifier.fillMaxWidth(),
                    singleLine = true
                )
                Spacer(modifier = Modifier.height(8.dp))
                OutlinedTextField(
                    value = requestMethod,
                    onValueChange = { requestMethod = it.uppercase() },
                    label = { Text("Request Method") },
                    supportingText = { Text("GET, POST, PUT, PATCH, DELETE, or HEAD") },
                    modifier = Modifier.fillMaxWidth(),
                    singleLine = true
                )
                Spacer(modifier = Modifier.height(8.dp))
                OutlinedTextField(
                    value = requestUrl,
                    onValueChange = { requestUrl = it },
                    label = { Text("Request URL") },
                    modifier = Modifier.fillMaxWidth()
                )
                Spacer(modifier = Modifier.height(8.dp))
                OutlinedTextField(
                    value = bodyType,
                    onValueChange = { bodyType = it },
                    label = { Text("Body Type") },
                    supportingText = { Text("None, MultipartFormData, FormURLEncoded, JSON, XML, or Binary") },
                    modifier = Modifier.fillMaxWidth(),
                    singleLine = true
                )
                Spacer(modifier = Modifier.height(8.dp))
                OutlinedTextField(
                    value = fileFormName,
                    onValueChange = { fileFormName = it },
                    label = { Text("File form name") },
                    modifier = Modifier.fillMaxWidth()
                )
                Spacer(modifier = Modifier.height(8.dp))
                OutlinedTextField(
                    value = parametersText,
                    onValueChange = { parametersText = it },
                    label = { Text("Parameters") },
                    supportingText = { Text("key=value, one per line") },
                    minLines = 3,
                    modifier = Modifier.fillMaxWidth()
                )
                Spacer(modifier = Modifier.height(8.dp))
                OutlinedTextField(
                    value = headersText,
                    onValueChange = { headersText = it },
                    label = { Text("Headers") },
                    supportingText = { Text("Header=Value, one per line") },
                    minLines = 3,
                    modifier = Modifier.fillMaxWidth()
                )
                Spacer(modifier = Modifier.height(8.dp))
                OutlinedTextField(
                    value = argumentsText,
                    onValueChange = { argumentsText = it },
                    label = { Text("Arguments") },
                    supportingText = { Text("field=value, one per line") },
                    minLines = 3,
                    modifier = Modifier.fillMaxWidth()
                )
                Spacer(modifier = Modifier.height(8.dp))
                OutlinedTextField(
                    value = dataText,
                    onValueChange = { dataText = it },
                    label = { Text("Body Data") },
                    minLines = 4,
                    modifier = Modifier.fillMaxWidth()
                )
                Spacer(modifier = Modifier.height(8.dp))
                OutlinedTextField(
                    value = urlText,
                    onValueChange = { urlText = it },
                    label = { Text("URL template") },
                    minLines = 2,
                    modifier = Modifier.fillMaxWidth()
                )
                Spacer(modifier = Modifier.height(8.dp))
                OutlinedTextField(
                    value = deletionUrlText,
                    onValueChange = { deletionUrlText = it },
                    label = { Text("Deletion URL template") },
                    minLines = 2,
                    modifier = Modifier.fillMaxWidth()
                )
                Spacer(modifier = Modifier.height(8.dp))
                OutlinedTextField(
                    value = errorMessageText,
                    onValueChange = { errorMessageText = it },
                    label = { Text("Error message template") },
                    minLines = 2,
                    modifier = Modifier.fillMaxWidth()
                )
            }
        },
        confirmButton = {
            Button(
                onClick = {
                    onSave(
                        entry.copy(
                            name = name.trim(),
                            destinationType = destinationType.trim(),
                            requestMethod = CustomUploaderRequestMethod.entries.firstOrNull {
                                it.name.equals(requestMethod.trim(), ignoreCase = true)
                            } ?: entry.requestMethod,
                            requestUrl = requestUrl.trim(),
                            bodyType = CustomUploaderBodyType.entries.firstOrNull {
                                it.name.equals(bodyType.trim(), ignoreCase = true)
                            } ?: entry.bodyType,
                            fileFormName = fileFormName.ifBlank { "file" },
                            parameters = KeyValueCodec.decode(parametersText),
                            headers = KeyValueCodec.decode(headersText),
                            arguments = KeyValueCodec.decode(argumentsText),
                            data = dataText,
                            url = urlText.trim(),
                            deletionUrl = deletionUrlText.trim(),
                            errorMessage = errorMessageText.trim(),
                            body = ""
                        )
                    )
                }
            ) {
                Text("Save")
            }
        },
        dismissButton = {
            OutlinedButton(onClick = onDismiss) { Text("Cancel") }
        }
    )
}
